src/util_base64.cpp

```diff
@@ -0,0 +1,489 @@
+  if (B3rem == 0) {
+    B3full = B3full - 1;
+    B3rem = 3;
+  }
+  // If the padding is mandatory, the remainder shall be encoded to a complete
+  // char[4] sequence.
+  const auto last_enc_len = cancel_pad ? (B3rem + 1) : 4;
+  const auto encoded_len = (B3full * 4) + last_enc_len;
+  // If requested the only number of required bytes for encoding, return it.
+  if ((nullptr == dst) && (0 == dst_size)) return encoded_len;
+  // Insufficient memory test.
+  if (dst_size < encoded_len) return 0;
+
+  // Main encode loop:
+  char last_enc4[4];
+  uint8_t last_bin3[3] = { 0, 0, 0 };
+  memcpy(&last_bin3[0], &src[B3full * 3], B3rem);
+  // First, process the complemented remainder.
+  const uint8_t *src_ = &last_bin3[0];
+  auto dst_ = &last_enc4[0];
+  for (size_t k = 0; k < B3full + 1; k++) {
+    dst_[0] = b64_tbl[src_[0] >> 2];
+    dst_[1] = b64_tbl[(0x3f & (src_[0] << 4)) | (src_[1] >> 4)];
+    dst_[2] = b64_tbl[(0x3f & (src_[1] << 2)) | (src_[2] >> 6)];
+    dst_[3] = b64_tbl[0x3f & src_[2]];
+    src_ = src + (k * 3);
+    dst_ = dst + (k * 4);
+  }
+  // Set padding ({a,_,_}=>{'x','y','=','='}, {a,b,_}=>{'x','y','z','='}).
+  if (B3rem < 3) last_enc4[3] = '=';
+  if (B3rem < 2) last_enc4[2] = '=';
+  // Copy remainder.
+  memcpy(&dst[B3full * 4], &last_enc4[0], last_enc_len);
+  return encoded_len;
+}
+
+// Returns the number of bytes, need to save a decoded result.
+// Return zero if an error or if src_size==0.
+// Error condition : (src_size != 0 && ret == 0).
+static size_t Base64DecodedSize(Base64Mode base64_mode, const char *src,
+                                size_t src_size, size_t *error_position) {
+  size_t err_pos = 0;
+  auto rc = B64DecodeImpl(base64_mode, src, src_size, nullptr, 0, &err_pos);
+  if (error_position) *error_position = err_pos;
+  return rc;
+}
+
+// Decodes the input base64 string into the binary sequence of bytes.
+// Expects: dst_size >= Base64DecodedSize(mode, src, src_size).
+// On success, returns the number of bytes written to the destination memory.
+// Returned Zero indicates nothing was written.
+// Error condition : (src_size != 0 && ret == 0).
+static size_t Base64Decode(Base64Mode base64_mode, const char *src,
+                           size_t src_size, uint8_t *dst, size_t dst_size,
+                           size_t *error_position) {
+  size_t err_pos = 0;
+  size_t rc = 0;
+  // If (0==dst_size) && (nullptr==src) then decode_b64 return the required
+  // size. Avoid this by check (dst_size>0).
+  if (dst_size > 0) {
+    rc = B64DecodeImpl(base64_mode, src, src_size, dst, dst_size, &err_pos);
+  }
+  if (error_position) *error_position = err_pos;
+  return rc;
+}
+
+// Encodes the input byte sequence into the base64 encoded string.
+static void Base64Encode(Base64Mode base64_mode, const uint8_t *src,
+                         size_t src_size, std::string *_text) {
+  const auto req_size = B64EncodeImpl(base64_mode, src, src_size, nullptr, 0);
```

Comment:
all the complexity in this function is unnecessary. a `std::string` can have variable amount of characters added. So remove this code for computing the size, and inline this function with `B64EncodeImpl` and simplify. Same for decode. This is the third time I am asking. I will stop my review here. Please do not ask me to review without addressing comments.